system/jlib/jevent.cpp

```diff
@@ -1754,6 +1758,913 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+        if (!available)
+            return visitSkipFile;
+        return visitContinue;
+    }
+
+    virtual Continuation visitAttribute(EventAttr id, const char* value) override
+    {
+        switch (id)
+        {
+        case EvAttrRecordedOption:
+            if (streq(value, "traceid"))
+                available = true;
+            break;
+        case EvAttrEventTraceId:
+            if (!streq(accepted, value))
+                return visitSkipEvent;
+            break;
+        default:
+            break;
+        }
+        return visitContinue;
+    }
+
+public:
+    CTraceIdFilterTerm(const char* traceId)
+        : accepted(traceId)
+    {
+    }
+
+protected:
+    StringAttr accepted;
+    bool available{false};
+};
+
+// Filter term that rejects all events for thread IDs other than the one given. If, during the
+// first event visit, it is observed that the file header did not identify thread IDs as recorded
+// data, the filter will reject the remainder of the file.
+class CThreadIdFilterTerm : public CEventVisitFilterTerm
+{
+public:
+    virtual bool visitFile(const char* filename, uint32_t version) override
+    {
+        available = false;
+        return true;
+    }
+
+    virtual Continuation visitEvent(EventType type) override
+    {
+        if (!available)
+            return visitSkipFile;
+        return visitContinue;
+    }
+
+    virtual Continuation visitAttribute(EventAttr id, const char* value) override
+    {
+        switch (id)
+        {
+        case EvAttrRecordedOption:
+            if (streq(value, "traceid"))
+                available = true;
+            break;
+        default:
+            break;
+        }
+        return visitContinue;
+    }
+
+    virtual Continuation visitAttribute(EventAttr id, __uint64 value) override
+    {
+        switch (id)
+        {
+        case EvAttrEventThreadId:
+            if (value != accepted)
+                return visitSkipEvent;
+            break;
+        default:
+            break;
+        }
+        return visitContinue;
+    }
+
+public:
+    CThreadIdFilterTerm(__uint64 threadId)
+        : accepted(threadId)
+    {
+    }
+
+protected:
+    __uint64 accepted;
+    bool available{false};
+};
+
+class CTimestampFilterTerm : public CEventVisitFilterTerm
+{
+public:
+    virtual Continuation visitAttribute(EventAttr id, __uint64 value) override
+    {
+        switch (id)
+        {
+        case EvAttrRecordedTimestamp:
+            {
+                __uint64 micros = (base + value ) / 1000;
+                // if ((base + value) % 1000)
+                //     micros++;
+                CDateTime recordingStart;
+                recordingStart.setTimeStamp(micros);
+                if (beforeEnd && recordingStart > end)
+                    return visitSkipFile;
+                base = value;
+                break;
+            }
+        case EvAttrEventTimeOffset:
+            {
+                __uint64 micros = (base + value ) / 1000;
```

Comment:
This should be implemented as a simple range test (similar to any other numeric field).

The values in the list should be adjusted when the EvAttrRecordedTimestamp field is seen.  As an alternative I think on balance the timestamp at the start of the file should be added to all time offsets, and passed through to the visitor as a real timestamp.  (Really the fact the an offset is stored in the file is an implementation detail that I don't think should be visible to the outside world).
That then means this filter becomes a simple numeric range filter.